crates/vqueues/src/scheduler/queue.rs

```diff
@@ -0,0 +1,264 @@
+    }
+
+    pub fn remove(&mut self, item_hash: u64) -> bool {
+        // Can this be the known head?
+        // Yes. Perhaps it expired/ended externally.
+        if let Head::Known(ref item) = self.head
+            && item.unique_hash() == item_hash
+        {
+            self.head = Head::Unknown;
+            // Ensure that next advance would re-seek to the newly added item
+            self.reader = Reader::Closed;
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Returns true if the head was changed
+    pub fn enqueue(&mut self, item: &S::Item) -> bool {
+        match (&self.head, &self.reader) {
+            // we are only unknown if we are new and didn't read the running list yet,
+            // we might also be in a limbo state if advance() failed.
+            (_, Reader::New { .. } | Reader::Running { .. }) => { /* do nothing */ }
+            (Head::Unknown, _) => { /* do nothing */ }
+            (Head::Empty, _) => {
+                self.reader = Reader::Closed;
+                self.head = Head::Known(item.clone());
+                return true;
+            }
+            (Head::Known(current), Reader::Inbox(_) | Reader::Closed) => {
+                if item < current {
+                    self.head = Head::Known(item.clone());
+                    // Ensure that next advance would re-seek to the newly added item
+                    self.reader = Reader::Closed;
+                    return true;
+                }
+            }
+        }
+        false
+    }
+
+    /// Returns the head if known, or None if the queue needs advancing
+    pub fn head(&self) -> Option<QueueItem<&S::Item>> {
+        match (&self.head, &self.reader) {
+            (Head::Unknown, _) => None,
+            (_, Reader::New { .. }) => None,
+            (Head::Known(item), Reader::Running { .. }) => Some(QueueItem::Running(item)),
+            (Head::Known(item), Reader::Inbox(_) | Reader::Closed) => Some(QueueItem::Inbox(item)),
+            (Head::Empty, _) => Some(QueueItem::None),
+        }
+    }
+
+    pub fn advance_if_needed(
+        &mut self,
+        storage: &S,
+        skip: &HashSet<u64>,
+        qid: &VQueueId,
+    ) -> Result<QueueItem<&S::Item>, StorageError> {
+        // Keep advancing until the head is known
+        while matches!(self.head, Head::Unknown) {
+            self.advance(storage, skip, qid)?;
+        }
+
+        match (&self.head, &self.reader) {
+            (Head::Unknown, _) => unreachable!("head must be known"),
+            (_, Reader::New { .. }) => unreachable!("reader cannot be new after poll"),
+            (Head::Known(item), Reader::Running { .. }) => Ok(QueueItem::Running(item)),
+            (Head::Known(item), Reader::Inbox(_) | Reader::Closed) => Ok(QueueItem::Inbox(item)),
+            (Head::Empty, _) => Ok(QueueItem::None),
+        }
+    }
+
+    pub fn advance(
+        &mut self,
+        storage: &S,
+        skip: &HashSet<u64>,
```

Comment:
The assumption is that `skip` will never contain `running` items, right? Maybe we could add a comment stating it.